wordpress-mobile/WordPress-Android
# Trigger Migration Flow when launching the app from deep links

Fixes #17653

This PR implements the logic to:
- Start the migration flow when the app is launched from deep links
- Land on the screen from the deep link after migration

> **Note** **Known issue**: When the Jetpack migration flow is started from a deep link, the notification permissions dialog is not shown on first launch of a freshly installed app. When the app is force closed and opened the second time, the dialog to enable notifications appears. Since the same behavior was already reproducible in the WordPress app this is not a blocker imho.

## To Test

#### Prerequisites & Setup

- ⚠️ Uninstall all versions of the WordPress and Jetpack apps from the device (including pre-alphas).
- ↗️ Open the [**links helper webpage**](https://codepen.io/ovitrif/full/WNKRBeV) in the device web browser & keep the browser open on this page.
- Clearing app data is required multiple times during testing, see **🆕 How to clear app data?** panel for steps.
- Setting which app to handle links is also needed, see **☑️ How to set app to handle verified links?** panel.

<details>
<summary><b>🆕 How to clear app data?</b></summary>

1. Open the apps drawer on the device
2. Tap & hold the icon of the desired app
3. Select `App Info` 
4. Tap `Storage` / `Storage & cache` (whichever presented)
5. Tap `Clear data` / `Clear storage` (whichever presented)
---
</details>

<details>
<summary><b>☑️ How to set app to handle verified links?</b></summary>

1. Open the apps drawer on the device
2. Tap & hold the icon of the desired app
3. Select `App Info` 
4. Tap `Open by default`
5. Tap `+ Add link`
6. ☑️ Check all three checkboxes then tap `Add`
---
</details>

#### 1. Jetpack App: Complete migration flow from `web link`

1. Build & Install WordPress app and log in (flavor: `wordpressWasabiDebug`)
2. Install Jetpack app of the same flavor & close it, then 🆕 clear app data
3. ☑️ Set Jetpack app to handle verified links
4. ↗️ Open `Stats` web link
5. **Expect** to see migration flow
7. Proceed to finish migration
8. **Expect** to land on the `Stats` screen
9. **Verify** your data was migrated

#### 2. Jetpack App: Complete migration flow from `deep link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ↗️ Open `Jetpack Reader` deep link with Jetpack app (just once)
4. **Expect** to see migration flow
5. Proceed to finish migration
6. **Expect** to land on the `Reader` screen

#### 3. Jetpack App: Complete migration flow from `post link`

1. **Prerequisite** WordPress app installed and logged in  (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links
4. ↗️ Open `Post` web link
5. **Expect** to see migration flow
6. Proceed to finish migration
7. **Expect** to land on `Reader` screen with the post `Post deep link test` open

#### 4. Jetpack App: Interrupt and resume migration from `web link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Notifications` web link
5. **Expect** to see migration flow
6. On the migration screen change device settings to `dark mode`
7. ~~Put the Jetpack app in background and bring it back to foreground~~ ([skip this step](https://github.com/wordpress-mobile/WordPress-Android/pull/17742#issuecomment-1381980835))
9. **Expect** The migration flow to continue
10. On the migration Welcome screen tap `continue`
11. Disable `dark theme`
12. **Expect** to remain on the migration Notifications screen
13. Proceed to finish migration
14. **Expect** To land on `Notifications` screen

#### 5. Jetpack App: Force close migration from `web link` and restart

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Stats` web link
5. **Expect** to see migration flow
6. Force close the Jetpack app
7. Reopen the Jetpack app
8. **Expect** to see migration flow
10. Proceed to finish migration
11. **Expect** to land on `My Site` screen
   **Note:** Even though it could be an option to land on the screen from the web link (`Stats`), this wouldn't be ideal because users could open the Jetpack app after a long time.

#### 6. Jetpack App: Exit migration and login from `web link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Notifications` web link
5. **Expect** to see migration flow
6. On the migration screen navigate back
7. **Expect** to see login screen
8. Login
9. **Expect** to land on the `My Site` screen
10. **Verify** that your data is not migrated (if you used the same account from WP app)

#### 7. Jetpack App: Open from `web link` without migration

1. 🆕 Clear WordPress app data
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Stats` web link
5. **Expect** to see login screen in Jetpack app
6. Login
7. **Expect** to land on the `Stats` screen

#### 8. WordPress App: Open `web link` in WordPress app then Jetpack via `Open links in Jetpack`

1. 🆕 Clear WordPress app data (should be the case)
2. Open WordPress app and login
3. 🆕 Clear Jetpack app data
4. ☑️ Set **WordPress** app to handle verified links
5. ↗️ Open `Notifications` web link
6. **Expect** to see `Open links in Jetpack?` screen
7. ☑️ Set **Jetpack** app to handle verified links
8. Return to WordPress app
9. Tap `Open links in Jetpack`
10. **Expect** to see migration flow
15. Proceed to finish migration
16. **Expect** to land on `Notifications` screen

#### 9. Regression: Jetpack App migration not eligible with Local Draft Content

1. **Prerequisite** WordPress app installed and logged in
2. Open the WordPress app and go to `Posts`
3. Disable internet connection on the device
4. Create a new blog post, add title and body text
5. Navigate `back` from the editor to save the post as local draft
6. Clear Jetpack app data
8. Open the Jetpack app
9. **Expect** to land on the login/ landing screen
10. Restore internet connection on the device
13. **Optionally** login and verify that login is successful

Eventually think of other regression testing scenarios that could occur during the migration flow and when opening the app from deep links and test them.

## Regression Notes
1. Potential unintended areas of impact
   Migration flow & deep link flows.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing & relied on existing migration and deep links tests.

3. What automated tests I added (or what prevented me from doing so)
   Added test for the updated logic in `JetpackAppMigrationFlowUtilsTest`.

PR submission checklist:
